Fix playback problems caused by missing audio packets

JHartman5 opened the pull request:
Attempts to solve https://github.com/dailymotion/hls.js/issues/409 by injecting 'silent' AAC-encoded audio frames when there are gaps in the middle of a segment. Helps to avoid nudges at the end of segments that include missing audio packets.

mangui commented:
thanks @JHartman5 
this PR share similarities with https://github.com/dailymotion/hls.js/pull/368
IMHO it would make more sense to inject silent AAC frames from within the MP4 remuxer rather than from the TS demuxer (which should stay workaround free as much as possible:TS demuxer could be used for other purpose than remuxing to MP4/ injection into MSE) .... so that all workarounds are located nearest to their needs (to accomodate web browsers, using fMP4)

JHartman5 commented:
I agree philosophically -- the silent frame injection should happen in the remuxer. I will try to refactor that into the remuxer tomorrow.

Do you think this will conflict with https://github.com/dailymotion/hls.js/pull/368? Do I need to take that into account when refactoring?

mangui commented:
yes, this will conflict with #368.
best would be to either rebase on #368 or take into account #368 functional need in your PR

mangui commented:
thanks for the update. i will check it ASAP
one note: it seems that mux.js is using different payload for silent frame depending on the sampling rate.
see https://github.com/videojs/mux.js/blob/f50426b257eb7d525311c1e6029e8891a2e07fe8/lib/data/silence.js

JHartman5 commented:
Thanks -- I commented on @imbcmdth's commit (https://github.com/videojs/mux.js/commit/f50426b257eb7d525311c1e6029e8891a2e07fe8) and asked him about how he generated his "cone of silence".

JHartman5 commented:
For the sake of completeness... I used GStreamer with a Ruby wrapper to generate my "cones of silence:

```ruby
(1..6).to_a.each do |num_channels|
  %w{main lc ltp}.each do |profile| # ssr doesn't seem to work...
    [8000, 11025, 12000, 16000, 22050, 24000, 32000, 44100, 48000, 64000, 88200, 96000].each do |sampling_rate|
      buffer_size = 1024 * 2 * num_channels
      output = `gst-launch-1.0 fakesrc blocksize=#{buffer_size} num-buffers=3 sizetype=fixed sizemax=#{buffer_size} filltype=zero ! "audio/x-raw, format=S16LE, layout=interleaved, rate=#{sampling_rate}, channels=#{num_channels}" ! faac ! "audio/mpeg, base-profile=(string)#{profile}, profile=(string)#{profile}" ! fakesink silent=false dump=true 2>/dev/null`
      results = output.scan /^([0-9a-fA-F]{8})[^:]*:((?: [0-9a-fA-F]{2})+)/
      buffers, current_buffer = [], "".b
      results.each do |r|
        address = r[0].to_i(16)
        data = r[1].strip.split(/\s/).collect {|b| b.to_i(16).chr.b}.join('')
        if address != 0
          if address == current_buffer.length
            current_buffer += data
          end
        else
          if !current_buffer.empty?
            buffers << current_buffer
            current_buffer = "".b
          end
          current_buffer += data
        end
      end
      buffers << current_buffer
      puts "sampling_rate=#{sampling_rate}, profile=#{profile}, num_channels=#{num_channels}, data=#{buffers[-1].split(//).collect {|c| '%02x' % c.ord}.join(' ')}"
    end
  end
end
```

mangui commented on src/remux/mp4-remuxer.js:
```diff
@@ -313,6 +319,38 @@ class MP4Remuxer {
+          ptsnorm0 = this._PTSNormalize(pts0, nextAacPts),
+          ptsnorm1 = this._PTSNormalize(pts1, nextAacPts),
+          missing = Math.round((ptsnorm1 - ptsnorm0) / pes2mp4ScaleFactor / 1024.0) - 1;
+      if (missing > 0) {
```
doing `PTS(i+1)-PTS(i)` to detect missing audio frames is useful to fix big holes, but it will not fix any PTS drift.

for example if suddenly 5 consecutive frames have PTS delta = 1500 (instead of 1024)
total drift will be (1500-1024)*5 = 2380 
ie : we will have a drift of 2 audio frames. but this won't be detected by this PTS delta check, as 
missing will always be equal to 0.

best would be to use `PTS(i+1)-1024*PTS(0)`   (scale factor omitted for sake of simplicity)

JHartman5 replied:
It wasn't _quite_ as easy to do as you suggested. But I did refactor this loop to basically quantize PTS values on 1024-sample boundaries, dropping or inserting frames as necessary to achieve that.

mangui commented on src/demux/adts.js:
```diff
@@ -6,6 +6,23 @@ import {ErrorTypes, ErrorDetails} from '../errors';
 
  class ADTS {
 
+  static getSilentFrame(channelCount) {
```
those silent frames are not related to ADTS, they are raw AAC frames.
I would suggest to move them to a new class remuxer/AACSilence

mangui commented on src/demux/tsdemuxer.js:
```diff
@@ -15,6 +15,7 @@
 // import Hex from '../utils/hex';
  import {logger} from '../utils/logger';
  import {ErrorTypes, ErrorDetails} from '../errors';
+ import '../utils/polyfill';
```
I guess polyfill needs to be imported from mp4-remuxer ?

mangui commented:
thanks for the update @JHartman5 , I will review it thoroughly on monday

mangui commented on src/helper/aac.js:
```diff
@@ -0,0 +1,24 @@
+ */
+
+class AAC {
+  static getSilentFrame(channelCount) {
```
Hi @JHartman5 I am just wondering how you generated those silent AAC frames.
do you have a command line to share ? I guess it is standard AAC ?
I am investigating https://github.com/dailymotion/hls.js/issues/868#issuecomment-268994938 and I am wondering whether we would need another set of silent audio frames for HE-AAC

mangui replied:
ok, found this
https://github.com/dailymotion/hls.js/pull/412#issuecomment-216220741
faac only produces AAC, so we would need to check the output with another audio encoder, like libfdk_aac

JHartman5 replied:
Yes, almost certainly. I'm sorry I didn't do that - I didn't have a convenient way to encode HE-AAC.

mangui replied:
no issue, I am struggling with setting up gstreamer.
it would be good to generate silent HE-AAC frame just to see how they look like

mangui replied:
i didn't succeed with gstreamer, I went back to ffmpeg
this cmd generates a 100ms 44100 Hz, AAC stereo silent audio frame

`ffmpeg -y -f lavfi -i "aevalsrc=0|0:d=0.1" -c:a libfaac output.aac && hexdump -C -s 33 -v output.aac`


```
00000021  ff f1 50 80 02 1f fc 21  00 49 90 02 19 00 23 80  |..P....!.I....#.|
00000031  ff f1 50 80 02 1f fc 21  00 49 90 02 19 00 23 80  |..P....!.I....#.|
00000041  ff f1 50 80 02 1f fc 21  00 49 90 02 19 00 23 80  |..P....!.I....#.|
00000051  ff f1 50 80 02 1f fc 21  00 49 90 02 19 00 23 80  |..P....!.I....#.|
00000061  ff f1 50 80 02 1f fc 21  00 49 90 02 19 00 23 80  |..P....!.I....#.|
```

this matches with
[[0x21, 0x00, 0x49, 0x90, 0x02, 0x19, 0x00, 0x23, 0x80]](https://github.com/dailymotion/hls.js/blob/master/src/helper/aac.js#L10)

but when trying to setup libfdk_aac

`ffmpeg -y -f lavfi -i "aevalsrc=0|0:d=0.1" -c:a libfdk_aac -profile:a aac_he_v2 output.aac && hexdump -C -s 33 -v output.aac
`

I got weird outputs

```
00000111  5a 5a 5a 5a 5a 5e ff f1  5c 40 22 e1 e8 01 40 22  |ZZZZZ^..\@"...@"|
00000121  80 a3 78 0e 81 00 00 00  00 00 e2 f5 22 ff 00 1f  |..x........."...|
00000131  fc 00 06 fe a1 0a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |......ZZZZZZZZZZ|
```

I would assume that the corresponding HE-AAC v2 silent audio frame is 
`01 40 22 80 a3 78 0e 81 00 00 00  00 00 e2 f5 22 ff 00 1f fc 00 06 fe a1`

but i am a little bit blind, so any help would be welcome to generate those frames

JHartman5 replied:
To me, it looks like a complete frame in the .aac file is:

`00000000  ff f1 58 40 22 e1 e8 01  40 22 80 a3 6e fb 60 9c  |..X@"...@"..n.`.|
00000010  08 00 00 00 00 00 12 a0  00 06 fe d1 0a 5a 5a 5a  |.............ZZZ|
00000020  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000030  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000040  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000050  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000060  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000070  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000080  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000090  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000a0  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000b0  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000c0  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000d0  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000e0  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000f0  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000100  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000110  5a 5a 5a 5a 5a 5a 5e                              |ZZZZZZ^|`

So if you removed the AAC ADTS header (7 bytes), you would be left with:

`00000007  01 40 22 80 a3 6e fb 60  9c 08 00 00 00 00 00 12  |.@"..n.`........|
00000017  a0 00 06 fe d1 0a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |......ZZZZZZZZZZ|
00000027  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000037  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000047  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000057  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000067  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000077  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000087  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000097  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000a7  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000b7  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000c7  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000d7  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000e7  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
000000f7  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5a  |ZZZZZZZZZZZZZZZZ|
00000107  5a 5a 5a 5a 5a 5a 5a 5a  5a 5a 5a 5a 5a 5a 5a 5e  |ZZZZZZZZZZZZZZZ^|`

I really, truly have no idea if that's valid HE-AAC or not, or if it varies by bitrate / number of channels, but I _think_ your approach seems correct.
